File: apps/android/feature/recording/src/main/kotlin/com/sigmundgranaas/turbo/expressive/feature/recording/TrackImport.kt
package com.sigmundgranaas.turbo.expressive.feature.recording

import com.sigmundgranaas.turbo.expressive.core.geo.GeoPath
import com.sigmundgranaas.turbo.expressive.core.geo.GeoPathSource
import com.sigmundgranaas.turbo.expressive.domain.LatLng
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/** A track parsed from an imported file: an optional name plus its geometry. */
data class ParsedTrack(val name: String?, val geo: GeoPath)

/**
 * Reads GPX, KML and GeoJSON track files into a [ParsedTrack], auto-detecting
 * the format from the content. Mirrors the export side ([serialize]); the parse
 * is pure so it can be unit-tested without a file picker. Returns null when no
 * usable line geometry is found.
 */
internal object TrackImport {
    private val json = Json { ignoreUnknownKeys = true }

    fun parse(body: String): ParsedTrack? {
        val trimmed = body.trimStart()
        return when {
            trimmed.startsWith("{") -> parseGeoJson(trimmed)
            trimmed.contains("<gpx", ignoreCase = true) -> parseGpx(body)
            trimmed.contains("<kml", ignoreCase = true) -> parseKml(body)
            else -> null
        }
    }

    private fun parseGpx(body: String): ParsedTrack? {
        // trkpt / rtept carry the line; wpt are standalone points we ignore for a track.
        val pointRegex = Regex(
            """<(?:trkpt|rtept)\b[^>]*\blat="([-\d.]+)"[^>]*\blon="([-\d.]+)"[^>]*?>(.*?)</(?:trkpt|rtept)>|""" +
                """<(?:trkpt|rtept)\b[^>]*\blat="([-\d.]+)"[^>]*\blon="([-\d.]+)"[^>]*/>""",
            RegexOption.DOT_MATCHES_ALL,
        )
        val points = mutableListOf<LatLng>()
        val elevations = mutableListOf<Double?>()
        pointRegex.findAll(body).forEach { m ->
            val lat = (m.groupValues[1].ifEmpty { m.groupValues[4] }).toDoubleOrNull() ?: return@forEach
            val lon = (m.groupValues[2].ifEmpty { m.groupValues[5] }).toDoubleOrNull() ?: return@forEach
            points += LatLng(lat, lon)
            elevations += Regex("""<ele>([-\d.]+)</ele>""").find(m.groupValues[3])?.groupValues?.get(1)?.toDoubleOrNull()
        }
        if (points.size < 2) return null
        return ParsedTrack(firstTag(body, "name"), geoOf(points, elevations))
    }

    private fun parseKml(body: String): ParsedTrack? {
        val points = mutableListOf<LatLng>()
        val elevations = mutableListOf<Double?>()
        Regex("""<coordinates>(.*?)</coordinates>""", RegexOption.DOT_MATCHES_ALL).findAll(body).forEach { block ->
            block.groupValues[1].trim().split(Regex("\\s+")).forEach { tuple ->
                val parts = tuple.split(",")
                val lon = parts.getOrNull(0)?.toDoubleOrNull()
                val lat = parts.getOrNull(1)?.toDoubleOrNull()
                if (lat != null && lon != null) {
                    points += LatLng(lat, lon)
                    elevations += parts.getOrNull(2)?.toDoubleOrNull()
                }
            }
        }
        if (points.size < 2) return null
        return ParsedTrack(firstTag(body, "name"), geoOf(points, elevations))
    }

    private fun parseGeoJson(body: String): ParsedTrack? {
        val root = runCatching { json.parseToJsonElement(body).jsonObject }.getOrNull() ?: return null
        // Unwrap FeatureCollection → first Feature, Feature → geometry.
        val name = root["features"]?.let { (it as? JsonArray) }
            ?.firstOrNull()?.jsonObject?.get("properties")?.jsonObject?.get("name")?.jsonPrimitive?.contentOrNull
            ?: root["properties"]?.jsonObject?.get("name")?.jsonPrimitive?.contentOrNull
        val geometry = findGeometry(root) ?: return null
        val coordsEl = geometry["coordinates"] ?: return null
        val type = geometry["type"]?.jsonPrimitive?.contentOrNull
        val lines = when (type) {
            "LineString" -> listOf(coordsEl as? JsonArray ?: return null)
            "MultiLineString" -> (coordsEl as? JsonArray)?.mapNotNull { it as? JsonArray } ?: return null
            else -> return null
        }
        val points = mutableListOf<LatLng>()
        val elevations = mutableListOf<Double?>()
        lines.forEach { line ->
            line.forEach { pair ->
                val arr = pair as? JsonArray ?: return@forEach
                val lon = (arr.getOrNull(0) as? JsonPrimitive)?.doubleOrNull
                val lat = (arr.getOrNull(1) as? JsonPrimitive)?.doubleOrNull
                if (lat != null && lon != null) {
                    points += LatLng(lat, lon)
                    elevations += (arr.getOrNull(2) as? JsonPrimitive)?.doubleOrNull
                }
            }
        }
        if (points.size < 2) return null
        return ParsedTrack(name, geoOf(points, elevations))
    }

    private fun findGeometry(root: JsonObject): JsonObject? = when {
        root["type"]?.jsonPrimitive?.contentOrNull == "FeatureCollection" ->
            (root["features"] as? JsonArray)?.firstOrNull()?.jsonObject?.get("geometry")?.jsonObject
        root["type"]?.jsonPrimitive?.contentOrNull == "Feature" -> root["geometry"]?.jsonObject
        root.containsKey("coordinates") -> root
        else -> null
    }

    private fun geoOf(points: List<LatLng>, elevations: List<Double?>): GeoPath {
        val hasEle = elevations.any { it != null }
        return GeoPath.fromPoints(points, GeoPathSource.Saved, if (hasEle) elevations else null)
    }

    private fun firstTag(body: String, tag: String): String? =
        Regex("<$tag>(.*?)</$tag>", RegexOption.DOT_MATCHES_ALL).find(body)
            ?.groupValues?.get(1)?.trim()?.takeIf { it.isNotEmpty() }
}

File: apps/web/src/map/paths/trackImport.ts
import { type LatLng, haversineMeters } from '../../geo';

/** A track parsed from a GPX / KML / GeoJSON file. Elevations are parallel to
 *  points (null where the file omitted one). */
export interface ParsedTrack {
  name?: string;
  points: LatLng[];
  elevations: (number | null)[];
}

const num = (s: string | null | undefined): number | null => {
  if (s == null || s.trim() === '') return null;
  const n = parseFloat(s);
  return Number.isFinite(n) ? n : null;
};

/** Auto-detect format and parse a track. Returns null if fewer than 2 points.
 *  Ports the Android `TrackImport.parse` tolerance (GPX trkpt/rtept, KML
 *  coordinates, GeoJSON LineString/MultiLineString). */
export function parseTrack(text: string): ParsedTrack | null {
  const t = text.trim();
  let parsed: ParsedTrack | null = null;
  if (t.startsWith('{') || t.startsWith('[')) parsed = parseGeoJson(t);
  else if (/<kml[\s>]/i.test(t)) parsed = parseKml(t);
  else if (/<gpx[\s>]/i.test(t)) parsed = parseGpx(t);
  else parsed = parseGpx(t) ?? parseKml(t) ?? parseGeoJson(t); // last-ditch sniff
  if (!parsed || parsed.points.length < 2) return null;
  return parsed;
}

function parseGpx(text: string): ParsedTrack | null {
  let doc: Document;
  try {
    doc = new DOMParser().parseFromString(text, 'application/xml');
  } catch {
    return null;
  }
  if (doc.querySelector('parsererror')) return null;
  // Prefer track points; fall back to route points.
  let nodes = Array.from(doc.getElementsByTagName('trkpt'));
  if (nodes.length === 0) nodes = Array.from(doc.getElementsByTagName('rtept'));
  if (nodes.length === 0) return null;
  const points: LatLng[] = [];
  const elevations: (number | null)[] = [];
  for (const n of nodes) {
    const lat = num(n.getAttribute('lat'));
    const lng = num(n.getAttribute('lon'));
    if (lat == null || lng == null) continue;
    points.push({ lat, lng });
    elevations.push(num(n.getElementsByTagName('ele')[0]?.textContent ?? null));
  }
  const name = doc.getElementsByTagName('name')[0]?.textContent?.trim() || undefined;
  return { name, points, elevations };
}

function parseKml(text: string): ParsedTrack | null {
  let doc: Document;
  try {
    doc = new DOMParser().parseFromString(text, 'application/xml');
  } catch {
    return null;
  }
  if (doc.querySelector('parsererror')) return null;
  const blocks = Array.from(doc.getElementsByTagName('coordinates'));
  if (blocks.length === 0) return null;
  const points: LatLng[] = [];
  const elevations: (number | null)[] = [];
  for (const block of blocks) {
    const raw = (block.textContent ?? '').trim();
    for (const tuple of raw.split(/\s+/)) {
      const parts = tuple.split(',');
      if (parts.length < 2) continue;
      const lng = num(parts[0]);
      const lat = num(parts[1]);
      if (lat == null || lng == null) continue;
      points.push({ lat, lng });
      elevations.push(num(parts[2] ?? null));
    }
  }
  const name = doc.getElementsByTagName('name')[0]?.textContent?.trim() || undefined;
  return { name, points, elevations };
}

function parseGeoJson(text: string): ParsedTrack | null {
  let json: unknown;
  try {
    json = JSON.parse(text);
  } catch {
    return null;
  }
  const points: LatLng[] = [];
  const elevations: (number | null)[] = [];
  let name: string | undefined;

  const pushLine = (coords: unknown) => {
    if (!Array.isArray(coords)) return;
    for (const c of coords) {
      if (!Array.isArray(c) || c.length < 2) continue;
      const lng = num(String(c[0]));
      const lat = num(String(c[1]));
      if (lat == null || lng == null) continue;
      points.push({ lat, lng });
      elevations.push(c.length > 2 ? num(String(c[2])) : null);
    }
  };
  const handleGeometry = (g: { type?: string; coordinates?: unknown } | null | undefined) => {
    if (!g) return;
    if (g.type === 'LineString') pushLine(g.coordinates);
    else if (g.type === 'MultiLineString' && Array.isArray(g.coordinates)) g.coordinates.forEach(pushLine);
  };

  const root = json as { type?: string; geometry?: { type?: string; coordinates?: unknown }; properties?: { name?: string }; features?: unknown[]; coordinates?: unknown };
  if (root.type === 'FeatureCollection' && Array.isArray(root.features)) {
    for (const f of root.features) {
      const feat = f as { geometry?: { type?: string; coordinates?: unknown }; properties?: { name?: string } };
      handleGeometry(feat.geometry);
      if (!name) name = feat.properties?.name;
    }
  } else if (root.type === 'Feature') {
    handleGeometry(root.geometry);
    name = root.properties?.name;
  } else if (root.type === 'LineString' || root.type === 'MultiLineString') {
    handleGeometry({ type: root.type, coordinates: root.coordinates });
  }
  return { name, points, elevations };
}

/** Distance (metres) + ascent/descent from elevation deltas. Mirrors Android
 *  `GeoMetrics`. Elevation totals are undefined when no elevation data exists. */
export function trackStats(points: LatLng[], elevations: (number | null)[]): {
  distanceM: number;
  ascentM?: number;
  descentM?: number;
} {
  let distanceM = 0;
  for (let i = 1; i < points.length; i++) distanceM += haversineMeters(points[i - 1], points[i]);

  const hasEle = elevations.some((e) => e != null);
  if (!hasEle) return { distanceM };
  let ascentM = 0;
  let descentM = 0;
  let prev: number | null = null;
  for (const e of elevations) {
    if (e == null) continue;
    if (prev != null) {
      const d = e - prev;
      if (d > 0) ascentM += d;
      else descentM -= d;
    }
    prev = e;
  }
  return { distanceM, ascentM, descentM };
}
